codec/unknownproto/unknown_fields.go

```diff
@@ -0,0 +1,359 @@
+package unknownproto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+// CheckMismatchedFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the unknown fields by tagNumber and wireType, or mismatched wireTypes.
+func CheckMismatchedFields(b []byte, msg proto.Message) error {
+	if len(b) == 0 {
+		return nil
+	}
+
+	desc, ok := msg.(descriptorIface)
+	if !ok {
+		return fmt.Errorf("%T does not have a Descriptor() method", msg)
+	}
+
+	fieldDescProtoFromTagNum, _, err := getDescriptorInfo(desc, msg)
+	if err != nil {
+		return err
+	}
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("invalid length")
+		}
+
+		fieldDescProto, ok := fieldDescProtoFromTagNum[int32(tagNum)]
+		switch {
+		case ok:
+			// Assert that the wireTypes match.
+			if !canEncodeType(wireType, fieldDescProto.GetType()) {
+				return &errMismatchedWireType{
+					Type:         reflect.ValueOf(msg).Type().String(),
+					TagNum:       tagNum,
+					GotWireType:  wireType,
+					WantWireType: protowire.Type(fieldDescProto.WireType()),
+				}
+			}
+
+		default:
+			if tagNum&bit11NonCritical == 0 {
```

Comment:
I think we need a boolean flag to enable non critical fields. We actually can't enable this while amino JSON signing is being used.

Reply:
No biggie, I've added a Checker type, whose AllowUnknownCriticalFields field by default is off. It has to be consciously opted into to now allow non-critical types. By default invoking the package level `CheckMismatchedFields` function will report every single unknown field, and I've provided guidance in the godoc.